Widen the user-element matrix from 60 to 765 DOF.

The F-barES-FEM-T4 volumetric element U3 spans the support of E A^c, up to
173 nodes (519 DOF) at c = 1.  The stock user-element path declares s(60,60),
enough only for the largest stock element (C3D20).  This patch gives the
user-element path its own arrays sized to the encoding limit (255 nodes,
765 DOF), so U3 at c = 1 fits.

It also adds the two arguments `lakon, ne` to the e_c3d_u call sites, which
the F-barES-FEM-T4 elements need to walk their base-U4 node->element map
(see the 0002 element-wiring patch): without them every argument after
position 14 in e_c3d_u shifts by two, so ntmat_ arrives NULL and the
assembly segfaults on the first U2 element.

--- mafillsm.f
+++ mafillsm.f
@@ -78,7 +78,29 @@
      &     time,thicke(mi(3),*),doubleglob(*),clearini(3,9,*),ttime,
      &     pslavsurf(3,*),pmastsurf(6,*),smscale(*),aut(*),val,
      &     aute(96)
+!
+!     F-barES-FEM-T4: wide element matrix for the user-element path.  The F-bar
+!     volumetric element U3 spans the support of E A^c, so its DOF count
+!     is set by mesh valence, not by a fixed node count: 30-40 nodes is
+!     routine in an unstructured tet mesh.  The stock s(60,60) holds 20
+!     nodes and was being overrun silently.  Standard elements keep using
+!     s/sm/ff.
+!
+      integer iuel,nduser
+      parameter(nduser=765)
+      real*8 ffu(nduser),sval,smval,ffval
+!
+!     su/smu MUST be allocatable, not fixed-size locals.  gfortran
+!     puts a local array over ~32 kB in static storage (.bss), and
+!     ccx assembles with pthreads -- so a fixed su(150,150) is one
+!     array SHARED by every assembly thread.  That is why the stock
+!     s(60,60) is safe at 28.8 kB and a widened copy is not.
+!     Allocatable locals live on the heap per invocation and are
+!     released automatically on return.
+!
+      real*8, allocatable :: su(:,:),smu(:,:)
 !     
+      allocate(su(nduser,nduser),smu(nduser,nduser))
       kflag=2
       i0=0
       icalccg=0
@@ -165,6 +187,16 @@
             nope=1
             ndof=3
           elseif(lakon(i)(1:1).eq.'U') then
+!
+!     F-barES-FEM-T4: U3, the F-barES-FEM-T4 volumetric edge domain, has a
+!     NON-SYMMETRIC element matrix (eq. 17).  The assembly below walks only
+!     the upper triangle (do ll=jj,...) into symmetric storage, which would
+!     silently symmetrise it.  When nasym is set -- linstatic/nonlingeo set
+!     it from the presence of U3, before this routine runs -- mafillsmas
+!     assembles U3 in full instead.  Same arrangement as friction contact
+!     ('ES...C') a few lines up.
+!
+            if((lakon(i)(1:2).eq.'U3').and.(nasym.eq.1)) cycle
             ndof=ichar(lakon(i)(7:7))
             nope=ichar(lakon(i)(8:8))
           else
@@ -283,6 +315,7 @@
 !     mortar end
 !     
           if(lakon(i)(1:1).ne.'U') then
+            iuel=0
             call e_c3d(co,kon,lakon(i),p1,p2,om,bodyf,nbody,s,sm,ff,i,
      &           nmethod,elcon,nelcon,rhcon,nrhcon,alcon,nalcon,
      &           alzero,ielmat,ielorien,norien,orab,ntmat_,
@@ -299,7 +332,10 @@
      &           set,nset,islavquadel,aute,irowte,jqte,
      &           mortartrafoflag)
           else
-            call e_c3d_u(co,kon,lakon(i),p1,p2,om,bodyf,nbody,s,sm,ff,i,
+            iuel=1
+            call e_c3d_u(co,kon,lakon(i),p1,p2,om,bodyf,nbody,
+     &           su,smu,ffu,i,
+     &           lakon,ne,
      &           nmethod,elcon,nelcon,rhcon,nrhcon,alcon,nalcon,
      &           alzero,ielmat,ielorien,norien,orab,ntmat_,
      &           t0,t1,ithermal,vold,iperturb,nelemload,sideload,xload,
@@ -375,9 +411,19 @@
 !     
               node2=kon(indexe+l)
               jdof2=nactdof(m,node2)
+!
+!     F-barES-FEM-T4: user elements were assembled into the wide su/smu.
+!
+              if(iuel.eq.1) then
+                sval=su(jj,ll)
+                smval=smu(jj,ll)
+              else
+                sval=s(jj,ll)
+                smval=sm(jj,ll)
+              endif
 c              if(i.eq.96) then
-c                write(20,100) node1,k,node2,m,s(jj,ll)
-c                write(21,100) node1,k,node2,m,sm(jj,ll)
+c                write(20,100) node1,k,node2,m,sval
+c                write(21,100) node1,k,node2,m,smval
 c              endif
 c 100          format(i10,",",i5,",",i10,",",i5,",",e20.13)
 !     
@@ -386,10 +432,10 @@
               if((jdof1.gt.0).and.(jdof2.gt.0)) then
                 if(stiffonly(1).eq.1) then
                   call add_sm_st(au,ad,jq,irow,jdof1,jdof2,
-     &                 s(jj,ll),jj,ll)
+     &                 sval,jj,ll)
                 else
                   call add_sm_ei(au,ad,aub,adb,jq,irow,jdof1,jdof2,
-     &                 s(jj,ll),sm(jj,ll),jj,ll)
+     &                 sval,smval,jj,ll)
                 endif
               elseif((jdof1.gt.0).or.(jdof2.gt.0)) then
 !     
@@ -414,14 +460,14 @@
                     if(index.eq.0) cycle
                     do
                       idof2=nactdof(nodempc(2,index),nodempc(1,index))
-                      value=-coefmpc(index)*s(jj,ll)/coefmpc(ist)
+                      value=-coefmpc(index)*sval/coefmpc(ist)
                       if(idof1.eq.idof2) value=2.d0*value
                       if(idof2.gt.0) then
                         if(stiffonly(1).eq.1) then
                           call add_sm_st(au,ad,jq,irow,idof1,
      &                         idof2,value,i0,i0)
                         else
-                          valu2=-coefmpc(index)*sm(jj,ll)/
+                          valu2=-coefmpc(index)*smval/
      &                         coefmpc(ist)
 !     
                           if(idof1.eq.idof2) valu2=2.d0*valu2
@@ -446,7 +492,7 @@
 !     
                 if(rhsi.eq.1) then
                 elseif(nmethod.eq.2) then
-                  value=s(jj,ll)
+                  value=sval
                   icolumn=neq(2)-idof2/2
                   call add_bo_st(au,jq,irow,idof1,icolumn,value)
                 endif
@@ -477,14 +523,14 @@
                         idof2=nactdof(nodempc(2,index2),
      &                       nodempc(1,index2))
                         value=coefmpc(index1)*coefmpc(index2)*
-     &                       s(jj,ll)/coefmpc(ist)/coefmpc(ist)
+     &                       sval/coefmpc(ist)/coefmpc(ist)
                         if((idof1.gt.0).and.(idof2.gt.0)) then
                           if(stiffonly(1).eq.1) then
                             call add_sm_st(au,ad,jq,irow,
      &                           idof1,idof2,value,i0,i0)
                           else
                             valu2=coefmpc(index1)*coefmpc(index2)*
-     &                           sm(jj,ll)/coefmpc(ist)/coefmpc(ist)
+     &                           smval/coefmpc(ist)/coefmpc(ist)
                             call add_sm_ei(au,ad,aub,adb,jq,
      &                           irow,idof1,idof2,value,valu2,i0,i0)
                           endif
@@ -534,7 +580,7 @@
                         idof2=nactdof(nodempc(2,index2),
      &                       nodempc(1,index2))
                         value=coefmpc(index1)*coefmpc(index2)*
-     &                       s(jj,ll)/coefmpc(ist1)/coefmpc(ist2)
+     &                       sval/coefmpc(ist1)/coefmpc(ist2)
                         if(idof1.eq.idof2) value=2.d0*value
                         if((idof1.gt.0).and.(idof2.gt.0)) then
                           if(stiffonly(1).eq.1) then
@@ -542,7 +588,7 @@
      &                           idof1,idof2,value,i0,i0)
                           else
                             valu2=coefmpc(index1)*coefmpc(index2)*
-     &                           sm(jj,ll)/coefmpc(ist1)/coefmpc(ist2)
+     &                           smval/coefmpc(ist1)/coefmpc(ist2)
 !     
                             if(idof1.eq.idof2) valu2=2.d0*valu2
 !     
@@ -576,6 +622,11 @@
               endif
             enddo
 !     
+            if(iuel.eq.1) then
+              ffval=ffu(jj)
+            else
+              ffval=ff(jj)
+            endif
             if(rhsi.eq.1) then
 !     
 !     distributed forces
@@ -585,7 +636,7 @@
 !     updating the external force vector for dynamic
 !     calculations
 !     
-                if(nmethod.eq.4) fnext(k,node1)=fnext(k,node1)+ff(jj)
+                if(nmethod.eq.4) fnext(k,node1)=fnext(k,node1)+ffval
 !     
                 if(jdof1.le.0) then
                   if(nmpc.ne.0) then
@@ -599,7 +650,7 @@
      &                       nodempc(1,index))
                         if(jdof1.gt.0) then
                           fext(jdof1)=fext(jdof1)
-     &                         -coefmpc(index)*ff(jj)
+     &                         -coefmpc(index)*ffval
      &                         /coefmpc(ist)
                         endif
                         index=nodempc(3,index)
@@ -609,7 +660,7 @@
                   endif
                   cycle
                 endif
-                fext(jdof1)=fext(jdof1)+ff(jj)
+                fext(jdof1)=fext(jdof1)+ffval
               endif
             endif
 !     
--- mafilldmss.f
+++ mafilldmss.f
@@ -62,6 +62,7 @@
       real*8 co(3,*),coefmpc(*),xload(2,*),p1(3),smscale(1),dd,
      &     p2(3),ad(*),au(*),bodyf(3),xloadold(2,*),reltime,
      &     t0(*),t1(*),vold(0:mi(2),*),s(60,60),
+     &     ffu(765),
      &     ff(60),xl(3,20),voldl(0:mi(2),20),
      &     sti(6,mi(1),*),sm(60,60),stx(6,mi(1),*),
      &     elcon(0:ncmat_,ntmat_,*),rhcon(0:1,ntmat_,*),springarea(2,*),
@@ -73,7 +74,13 @@
      &     time,thicke(mi(3),*),doubleglob(*),clearini(3,9,*),damping,
      &     pslavsurf(3,*),pmastsurf(6,*),freq,stiff(21),dacon(*),
      &     aut(1)
+!
+!     allocatable, not a fixed local: gfortran would place a 180 kB
+!     local array in static storage, shared by every pthread.
+!
+      real*8, allocatable :: su(:,:),smu(:,:)
 !     
+      allocate(su(765,765),smu(765,765))
       mortartrafoflag=0
       one=1
 !     
@@ -295,8 +302,8 @@
      &           set,nset,islavquadel,
      &           aut,irowt,jqt,mortartrafoflag)
           else
-            call e_c3d_u(co,kon,lakon(i),p1,p2,om,bodyf,nbody,s,sm,
-     &           ff,i,nmethod,elcon,nelcon,rhcon,nrhcon,alcon,nalcon,
+            call e_c3d_u(co,kon,lakon(i),p1,p2,om,bodyf,nbody,su,smu,
+     &           ffu,i,lakon,ne,nmethod,elcon,nelcon,rhcon,nrhcon,alcon,nalcon,
      &           alzero,ielmat,ielorien,norien,orab,ntmat_,
      &           t0,t1,ithermal,vold,iperturb,nelemload,sideload,xload,
      &           nload,idist,sti,stx,iexpl,plicon,
@@ -308,6 +315,24 @@
      &           integerglob,doubleglob,tieset,istartset,
      &           iendset,ialset,ntie,nasym,
      &           ielprop,prop,nope)
+!
+!     F-barES-FEM-T4: this damping path assembles from the stock s/sm/ff, which
+!     hold 20 nodes.  A large U3 volumetric stencil cannot be represented
+!     here, so refuse rather than silently truncating it.
+!
+            if(nope.gt.20) then
+              write(*,*) '*ERROR in mafilldmss: user element ',i
+              write(*,*) '       spans ',nope,' nodes; this path'
+              write(*,*) '       holds 20.'
+              call exit(201)
+            endif
+            do jj=1,3*nope
+              ff(jj)=ffu(jj)
+              do ll=1,3*nope
+                s(jj,ll)=su(jj,ll)
+                sm(jj,ll)=smu(jj,ll)
+              enddo
+            enddo
           endif
           do jj=1,ndof*nope
             do ii=1,jj
